bn_mp_set_double.c does not compile under MacOS plus

Under MacOS...
```
$ gcc --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
```

a simple `make` returns...
```
   * cc bn_mp_set_double.o
bn_mp_set_double.c:55:6: warning: "mp_set_double implementation is only available on platforms with IEEE754 floating point format" [-W#warnings]
#    warning "mp_set_double implementation is only available on platforms with IEEE754 floating point format"
     ^
1 warning generated.
```
...resulting in no binary for mp_set_double().  (mp_get_double() compiles fine.)


---
#### Notes:

 * LTM and LTC will not crash if mp_get_double() and mp_set_double() are not compiled.  They are not called fm anywhere else in LTM or LTC.  ...and only mp_set_double(), demo.c, and mpi.c are conditioned with `#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559)`

  * Looking at the list of enabled defines in the attached file ( [bn_mp_set_double.i.txt](https://github.com/libtom/libtommath/files/2855908/bn_mp_set_double.i.txt) ), make note of those with FLT, DBL, and LDBL.

  * It appears when Apple's defines `__DBL_HAS_DENORM__`, `__DBL_HAS_INFINITY__`, and `__DBL_HAS_QUIET_NAN__` are true the processor supports IEEE 754/IEC 60559.  Perhaps some combination of these 3 defines could be used to set `__STDC_IEC_559__`? 

  * LTM only implements get/set for DBL.  Apple supports FLT and LDBL as well.  So, for the sake of completeness, perhaps we should also implement mp_get|set_float() and mp_get|set_longdouble()?

  * Finally, out of a concern for some misusing mp_get|set_double(), the doc should provide a strong caution about the possible loss of precision.

  * FWIW, all Macs since 2006 have Intel processors with integrated IEEE-754 compliant FPUs.  ...as do all PPC Macs.  M680x0 Mac FPUs varies.

Appveyor (windows):
```
<snip>

compiling 3rdparty\libtommath\bn_mp_set_double.obj
bn_mp_set_double.c
mp_set_double implementation is only available on platforms with IEEE754 floating point format

<snip>

linking moar.dll
   Creating library moar.dll.lib and object moar.dll.exp
bigintops.obj : error LNK2001: unresolved external symbol mp_set_double
moar.dll : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\link.EXE"' : return code '0x460'
Stop.
```
TravisCI (MacOS only, linux is fine):
```
<snip>

compiling 3rdparty/libtommath/bn_mp_set_double.o
3rdparty/libtommath/bn_mp_set_double.c:44:6: warning: "mp_set_double implementation is only available on platforms with IEEE754 floating point format" [-W#warnings]

#    warning "mp_set_double implementation is only available on platforms with IEEE754 floating point format"

     ^

1 warning generated.

<snip>

linking 3rdparty/libtommath/libtommath.a

/Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: 3rdparty/libtommath/libtommath.a(bn_mp_set_double.o) has no symbols

linking 3rdparty/libuv/libuv.a

linking libmoar.dylib

Undefined symbols for architecture x86_64:

  "_mp_set_double", referenced from:

      _MVM_bigint_from_num in bigintops.o

ld: symbol(s) not found for architecture x86_64

clang: error: linker command failed with exit code 1 (use -v to see invocation)

make: *** [libmoar.dylib] Error 1

The command "make -j2 install;" exited with 2.
```

--- a/demo/test.c
+++ b/demo/test.c
@@ -522,7 +522,7 @@ LBL_ERR:
 
 }
 
-#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559) || defined(__x86_64__) || defined(_M_X64) || defined(_M_AMD64) || defined(__i386__) || defined(_M_X86) || defined(__aarch64__) || defined(__arm__)
+#if defined(MP_HAS_SET_DOUBLE)
 
 #ifdef _MSC_VER
 #pragma warning(push)
@@ -2172,7 +2172,7 @@ static int unit_tests(int argc, char **argv)
       T1(mp_reduce_2k_l, MP_REDUCE_2K_L),
       T1(mp_radix_size, MP_RADIX_SIZE),
       T1(s_mp_radix_size_overestimate, S_MP_RADIX_SIZE_OVERESTIMATE),
-#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559)
+#if defined(MP_HAS_SET_DOUBLE)
       T1(mp_set_double, MP_SET_DOUBLE),
 #endif
       T1(mp_signed_rsh, MP_SIGNED_RSH),

--- a/demo/test.c
+++ b/demo/test.c
@@ -523,18 +523,23 @@ LBL_ERR:
 }
 
 #if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559) || defined(__x86_64__) || defined(_M_X64) || defined(_M_AMD64) || defined(__i386__) || defined(_M_X86) || defined(__aarch64__) || defined(__arm__)
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable: 4723) /* potential divide by 0 */
+#endif
 static int test_mp_set_double(void)
 {
    int i;
+   double dbl_zero = 0.0;
 
    mp_int a, b;
    DOR(mp_init_multi(&a, &b, NULL));
-
    /* test mp_get_double/mp_set_double */
-   EXPECT(mp_set_double(&a, +1.0/0.0) == MP_VAL);
-   EXPECT(mp_set_double(&a, -1.0/0.0) == MP_VAL);
-   EXPECT(mp_set_double(&a, +0.0/0.0) == MP_VAL);
-   EXPECT(mp_set_double(&a, -0.0/0.0) == MP_VAL);
+   EXPECT(mp_set_double(&a, +1.0/dbl_zero) == MP_VAL);
+   EXPECT(mp_set_double(&a, -1.0/dbl_zero) == MP_VAL);
+   EXPECT(mp_set_double(&a, +0.0/dbl_zero) == MP_VAL);
+   EXPECT(mp_set_double(&a, -0.0/dbl_zero) == MP_VAL);
 
    for (i = 0; i < 1000; ++i) {
       int tmp = rand_int();
@@ -552,6 +557,9 @@ LBL_ERR:
    return EXIT_FAILURE;
 
 }
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
 #endif
 
 static int test_mp_get_u32(void)

--- a/demo/test.c
+++ b/demo/test.c
@@ -522,7 +522,7 @@ LBL_ERR:
 
 }
 
-#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559)
+#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559) || defined(__x86_64__) || defined(_M_X64) || defined(_M_AMD64) || defined(__i386__) || defined(_M_X86) || defined(__aarch64__) || defined(__arm__)
 static int test_mp_set_double(void)
 {
    int i;

--- a/mp_set_double.c
+++ b/mp_set_double.c
@@ -3,7 +3,7 @@
 /* LibTomMath, multiple-precision integer library -- Tom St Denis */
 /* SPDX-License-Identifier: Unlicense */
 
-#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559)
+#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559) || defined(__x86_64__) || defined(_M_X64) || defined(_M_AMD64) || defined(__i386__) || defined(_M_X86) || defined(__aarch64__) || defined(__arm__)
 mp_err mp_set_double(mp_int *a, double b)
 {
    uint64_t frac;

--- a/mp_set_double.c
+++ b/mp_set_double.c
@@ -3,7 +3,7 @@
 /* LibTomMath, multiple-precision integer library -- Tom St Denis */
 /* SPDX-License-Identifier: Unlicense */
 
-#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559) || defined(__x86_64__) || defined(_M_X64) || defined(_M_AMD64) || defined(__i386__) || defined(_M_X86) || defined(__aarch64__) || defined(__arm__)
+#if defined(MP_HAS_SET_DOUBLE)
 mp_err mp_set_double(mp_int *a, double b)
 {
    uint64_t frac;

--- a/tommath_private.h
+++ b/tommath_private.h
@@ -164,6 +164,13 @@ MP_STATIC_ASSERT(prec_geq_min_prec, MP_DEFAULT_DIGIT_COUNT >= MP_MIN_DIGIT_COUNT
  */
 #define MP_MAX_DIGIT_COUNT ((INT_MAX - 2) / MP_DIGIT_BIT)
 
+#if defined(__STDC_IEC_559__) || defined(__GCC_IEC_559) \
+   || defined(__x86_64__) || defined(_M_X64) || defined(_M_AMD64) \
+   || defined(__i386__) || defined(_M_X86) \
+   || defined(__aarch64__) || defined(__arm__)
+#define MP_HAS_SET_DOUBLE
+#endif
+
 /* random number source */
 extern MP_PRIVATE mp_err(*s_mp_rand_source)(void *out, size_t size);
 

